Provisional KEP: Apply for client-go's typed clients (#2156)
kubernetes/enhancements: keps/sig-api-machinery/2144-clientgo-apply/README.md

-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

client-go's typed clients need a typesafe, programmatic way to make apply
requests.

## Motivation

Currently, the only way to invoke server side apply from client-go is to call
`Patch` with `PatchType.ApplyPatchType` and provide a `[]byte` containing the
YAML or JSON of the apply configuration. This has a couple important
deficiencies:

- It is a gap completeness of the type client, which provides typesafe APIs for
  all other major methods.
- It makes it to easy for developers to make a major, but non-obvious mistake:
  Use the existing go structs to construct an apply configuration, serialize
  it to JSON, and pass it to `Patch`. This can cause zero valued required
  fields being accientally included in the apply configuration resulting
  in fields being accidentally set to incorrect values and/or fields accidentally
  being accidentally being clamed as owned.

Both sig-api-machinery and wg-api-expression agree that this enhancement is

> required before server side apply to be promoted to GA.

Reviewer: ```suggestion
required for server side apply to be promoted to GA.
```

### Goals

Introduce a typesafe, programmatic way to call server side apply using the typed
client in client-go.

Express all apply configurations in Go that can be expressed in

> YAML. Specifically, an apply request must include only field that are set by

Reviewer: ```suggestion
YAML. Specifically, an apply request must include only fields that are set by
```

> applier and exclude those not set by applier.

Reviewer: ```suggestion
the applier and exclude those not set by the applier.
```

Validate this enhancement meets the needs of developers:

- An developer not directly involved in this enhancement successfully converts

>   a 1st party controller (one in github.com/kubernetes/kubernetes) to us this

Reviewer 2: ```suggestion
  a 1st party controller (one in github.com/kubernetes/kubernetes) to use this
```

  enhancement.
- A representative group of the developer community is made aware of this
  proposed enhancement, is given early access to it via a fork of
  controller-tools with the requisite generators, and is given the opportunity
  to try it out and provide feedback.

### Non-Goals

Enhancements to client-go's dynamic client. The client-go dynamic client already
supports Apply via Patch, which is adequate for the dynamic client.

Protobuf support. Apply does not support protobuf, and it will not be added with
this enhancement.

## Proposal

`Apply` functions should be included in the typed clients generated for
client-go and should accept the apply configuration using a strongly typed
representation, which will need to be generated for this purpose.

### Risks and Mitigations

#### Poor adoption

> Risk: Developers adoption is poor, either because the asthetics/ergonomics are

Reviewer: ```suggestion
Risk: Developers adoption is poor, either because the aesthetics/ergonomics are
```

not to their liking or the functinality is insufficient to do what they need to
do with it. This could lead to (a) poor server side apply adoption, and/or (b)
developers building alternate solutions.

Mitigation: We are working with the kubebuilder community to
get hands on feedback from developers to guide our design decisions around

> asthetics/egronomics with a goal of having both client-go and kubebuilder

Reviewer: ```suggestion
aesthetics/ergonomics with a goal of having both client-go and kubebuilder
```

take an aligned approach to adding apply to clients in a typesafe way.

## Design Details

### Apply functions

The client-go typed clients will be extended to include Apply functions, e.g.:

```go

> func (c *deployments) Apply(ctx Context, deployment *appsv1apply.Deployment, fieldManager string, metav1.ApplyOptions) (*Deployment, error)

Reviewer 2: super unfortunate to have fieldManager separate from apply options, any way we can combine / embed?

Reviewer 2: Last parameter is unnamed?

Reviewer 2: Maybe make it "otherOptions"?

Also what if you change the order to put the deployment object last, I feel like that will help readability if someone is going to inline the object definition?

Reviewer: I think it makes sense to have the fieldManager in the signature. Feels more explicit while things in options are optional.

Author: Fixed the last parameter. It is named, and there is also a `subresources ...string` parameter that I had accidentally omitted here. The signature is the same as `Patch` but with the `fieldManager` param required.

I'll move the fieldManager param to before the object. I'm hesitant to change the order of the params too much because there are parameter order conventions the method functions. For context:

```go
Create(ctx context.Context, deployment *v1.Deployment, opts metav1.CreateOptions) (*v1.Deployment, error)
Update(ctx context.Context, deployment *v1.Deployment, opts metav1.UpdateOptions) (*v1.Deployment, error)
UpdateStatus(ctx context.Context, deployment *v1.Deployment, opts metav1.UpdateOptions) (*v1.Deployment, error)
Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error
DeleteCollection(ctx context.Context, opts metav1.DeleteOptions, listOpts metav1.ListOptions) error
Get(ctx context.Context, name string, opts metav1.GetOptions) (*v1.Deployment, error)
List(ctx context.Context, opts metav1.ListOptions) (*v1.DeploymentList, error)
Watch(ctx context.Context, opts metav1.ListOptions) (watch.Interface, error)
Patch(ctx context.Context, name string, pt types.PatchType, data []byte, opts metav1.PatchOptions, subresources ...string) (result *v1.Deployment,  err error)
Apply(ctx context.Context, deployment *appsv1.DeploymentApplyConfiguration, fieldManager string, opts metav1.ApplyOptions, subresources ...string) (result *v1.Deployment, err error)
GetScale(ctx context.Context, deploymentName string, options metav1.GetOptions) (*autoscalingv1.Scale, error)
UpdateScale(ctx context.Context, deploymentName string, scale *autoscalingv1.Scale, opts metav1.UpdateOptions) (*autoscalingv1.Scale, error)
```

Author: I've added an `UNRESOLVED` to the KEP for how we organize the params.

> func (c *deployments) Apply(ctx Context, deployment *appsv1apply.Deployment, fieldManager string, metav1.ApplyOptions) (*Deployment, error)

Reviewer 2: ACTUALLY! Now that I've read the below, what about some variation of:

```
client.Deployments().BeginApply("field-manager-name", options).Body(
  appsv1apply.Deployment().
    SetObjectMeta(&metav1apply.ObjectMeta()). ...
).Apply(ctx)
```

Could shuffle the parameters around and change the method names. I don't know.

Author: I like how the options are clearly separated out. I'm a bit concerned about how different the Apply signature becomes from the other method functions (see the example I pasted in https://github.com/kubernetes/enhancements/pull/2156/files#r548100921).

Author: I've added an `UNRESOLVED` about we organize the options in the function signature so we make sure to address it fully before going to implementable.

```

`ApplyOptions` will be added to metav1 even though `PatchOptions` will continue
to be used over the wire. This will make it obvious in the Apply function

> signature that `fieldManager` is required.

Reviewer 2: Oh...

```go
type ApplyOptions struct {
  DryRun []string `json:"dryRun,omitempty" protobuf:"bytes,1,rep,name=dryRun"`

>   Force *bool `json:"force,omitempty" protobuf:"varint,2,opt,name=force"`

Reviewer 2: Are you sure you shouldn't make this a parameter too? User-like things want false and controller-like things want true, I don't think there's a great default?

Author: That's a really good point.

I've added an `UNRESOLVED` to this section for the `Force` field and for how we organize the params for Apply, since I feel like those both need more consideration.

}

func (ApplyOptions) ToPatchOptions(fieldManager string) PatchOptions
```

Apply will combine the `fieldManager` argument with `ApplyOptions` to create the
`PatchOptions`.

Each apply call will be required to provide a fieldmanager name. We will not

> provide a a way for the fieldmanager name to be set for the entire

Reviewer: ```suggestion
provide a way for the fieldmanager name to be set for the entire
```

clientset. There are a couple reasons for this:

- If a client has multiple code paths where it makes apply requests to the same
  object, but with different field sets, they must use different field manager
  names. If they use the same field manager name they will cause fields to be
  accidentally removed or disowned. This is a potential foot gun we would like to

>   avoid.

Reviewer: We should also document this. I don't know how to make this as clear as possible yet, but expect it to happen more than we like.
Like people just setting a fixed value/constant as their fieldManager and using it exactly like this.

Author: Created https://github.com/kubernetes/website/issues/25785 (with xref back to this comment) to track.

- Apply requests always conflict with update requests, even if they were made by
  the same client with the same field manager name. This is by design. So when a
  controller migrates from update to apply, it will need to deal with conflicts

>   regardless of what field manager name is used.

Reviewer 2: I don't understand what this has to do with forcing people to specify a name (the first point is sufficient to carry the argument).

Author: You're right. Removing.

### Generated apply configuration types

All fields present in an apply configuration become owned by the applier after
when the apply request succeeds. Go structs contain zero valued fields which are
included even if the user never explicitly sets the field. Required boolean
fields are a good example of fields that would be applied incorrectly using go

> structs, e.g. `ContainerStatus.Ready` (required, not omitempty). Because of this

Reviewer 3: I'm curious how much of a problem this is in practice. If I'm the first to apply an object, I have to specify the required fields anyways. If I'm applying on an existing object, I don't have to apply required fields, but are there scenarios, for example in the `containerStatus.Ready` example here where I would want to apply the `containerStatus` without knowing the ready state? I'd love to see a very convincing example in this doc!

Author: We got direct feedback about this in practice when we discussed this with the controller-runtime WG. One interesting thing was that multiple people chimed in reporting that they hadn't realized that there was a problem using the structs with apply until they got burned when a required boolean field got flipped to false during an apply that they had not intended to modify. Interestingly, it was specifically the boolean typed fields that they reported.

The `ContainerStatus.Ready` example makes a lot of sense to me. If it is already set to true, and you don't intend to set it to false when calling apply, it is really surprising to find that it was set to its default of false, which is exactly what can happen if using the existing go structs.

Author: Per our offline discussion I agree `ContainerStatus.Ready` is not an ideal example since it is difficult to imagine a usage pattern when it would not be included in the apply request.

I've rewritten this section to include more examples. I think `ContainerPort.ContainerPort` is a slightly better example, but anyone can review a larger list using a grep like:

```sh
find . -name "types.go" | xargs git grep -E " (int[0-9]{0,2}|bool) \`json:" | grep -v omitempty |grep -v vendor | grep -v "*"
```

I also did some searching and found that we have well over 100 fields like this, which I think adds weight to there being a fundamental soundness problem.

we cannot use the existing go structs to represent apply configurations.

<<[UNRESOLVED @jpbetz @jennybuckley ]>> 
Finalize which alternative to use based on developer feedback. See the
[Alternatives](#alternatives) for a complete list, but are currently focusing on
the two below alternatives. We are working with the Kubebuilder community to
gather feedback on what developers prefer.
<<[/UNRESOLVED]>>

> #### Alternative 1: Genreated structs where all fields are pointers

Reviewer: ```suggestion
#### Alternative 1: Generated structs where all fields are pointers
```

Example usage:

```go
&appsv1apply.Deployment{

>   Name: &pointer.StringPtr("nginx-deployment"),

Reviewer 3: I don't think you wanted to have a pointer to a pointer here (you can't take a pointer of an r-value).

Reviewer 2: +1

Author: Oops, this was a typo. The existing pointer package in Kubernetes that I'm assuming would be used here correctly returns a pointer so this wouldn't have compiled. Fixed.

  Spec: &appsv1apply.DeploymentSpec{

>     Replicas: &pointer.Int32Ptr(0),

Reviewer 2: If we really want people to do this, we probably want to make this package a "." import, or at least a really short name.

Author: Added a "ptr" import to the example.

    Template: &v1apply.PodTemplate{
      Spec: &v1apply.PodSpec{
        Containers: []v1.Containers{
          {
            Name: &pointer.StringPtr("nginx"),
            Image: &pointer.StringPtr("nginx:latest"),
          },
        }
      },
    },
  },
}
```

For this approach, developers need to use a library like
https://github.com/kubernetes/utils/blob/master/pointer/pointer.go
to inline primitive literals.

#### Alternative 2: Generated "builders"

Example usage:

```go
&appsv1apply.Deployment().

>   SetObjectMeta(&metav1apply.ObjectMeta().

Reviewer 2: Do you really need the "Set" prefix for everything?

Reviewer: Golang libraries uses `With` instead of `Set` a lot which I prefer. But Set is a known thing inside the Kubernetes types so far so feels more uniform.

Author: After a few discussions, I've switched the example to not prefix the builder name with "Set", so now we just have `ObjectMeta()` (instead of `SetObjectMeta()`). I've also added a `<<[UNRESOLVED` to this it doesn't block us from merging this KEP as provisional.

    SetName("nginx-deployment")).
  SetSpec(&appsv1apply.DeploymentSpec().
    SetReplicas(0).
    SetTemplate(
      &v1apply.PodTemplate().
        SetSpec(&v1apply.SetPodSpec().
          SetContainers(v1apply.ContainerList{
            v1apply.Container().
              SetName("nginx").
              SetImage("nginx:1.14.2")
            v1apply.Container().
              SetName("sidecar").

>           })

Reviewer 2: This looks way better to me than option 1.

I think ideal would be embedding yaml directly, as in the tests for smd? But... that's not typesafe and there's no way to make it typesafe. Rust macros would be awesome for that. :(

Author: I expect this to be the most contentious part of this KEP 😬 

I started strongly in support of this option as well. I'm now on the fence because quite a few engineers I've shown the alternatives to strongly favored option 1. Their main reason seems to be a strong preference toward the use of go structs, even if they need to deal with lots of pointers to literals. 

Let's merge as provisional, gather the feedback from the controller-runtime/kubebuilder community and then discuss in a sig-apimachinery bi-weekly.

The idea of using macros in a language that supports it is really interesting.

        )
      )
    )
  )
```

#### Comparison of alternatives

See https://github.com/kubernetes/kubernetes/pull/95988 for a working implementation
of alterative 1 and https://github.com/jpbetz/kubernetes/tree/apply-client-go-builders
for a working implementation of alternative 2.

Of the two leading alternatives--"builders" and "structs with pointers"--we implemented
prototypes of both. They had roughly equivalent performance, and no differences

> in their capabilities. The choice between the two is primarily one of asthetics/ergonomics.

Reviewer: ```suggestion
in their capabilities. The choice between the two is primarily one of aesthetics/ergonomics.
```

that will allow them to generate apply configuration types that have both the
alternatives. Our goal is to get feedback from developers after they try out the
generated apply configurations and use that to inform our decision.

While it is possible to generate types that have both the pointer fields exposed
and the builder functions, we would prefer to provide clear guidance to the
community on how apply configurations should be represented in go and encourage
consistent use of only one of these approaches.

#### DeepCopy support

If "structs with pointers" approach is used, the existing deepcopy-gen
can be used to generate deep copy impelemntations for the generated apply
configuration types.

#### Code Generator Changes

hack/update-codegen.sh and hack/verify-codegen.sh will be updated to generate
the apply functions and apply configuration types.

##### Addition of applyconfiguration-gen

- Add staging/src/k8s.io/code-generator/cmd/applyconfigurations-gen
- Generates into staging/vendor/k8s.io/client-go/applyconfigurations/
- Only generate builders for struct types reachable from the types that have the +clientgen annotation

> - Don't generate builders for MarshalJSON types (Quantity, IntOrString)

Reviewer 2: People construct these directly?

Author: You're right, that's a bad example (although they do implement MarshalJSON). Looks like Duration and Time are better examples, I'll use them here.

> - Don't generate builders for RawExtension or Unknown

Reviewer 2: How do people set fields of those types?

Author: I'm more familiar with RawExtension from admission webhooks: [declaration](https://github.com/kubernetes/kubernetes/blob/fec1a366c3b17b5d46b79ddac0b8bb04dfd212ee/staging/src/k8s.io/api/admission/v1/types.go#L98), [usage example](https://github.com/kubernetes/kubernetes/blob/fec1a366c3b17b5d46b79ddac0b8bb04dfd212ee/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/request/admissionreview_test.go#L536)

Here's what I found for Unknown:  [declaration](https://github.com/kubernetes/kubernetes/blob/fec1a366c3b17b5d46b79ddac0b8bb04dfd212ee/staging/src/k8s.io/apiserver/pkg/apis/audit/v1/types.go#L115), only usage I could find for this is in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/audit_test.go

Author: I've added the RawExtension example to the KEP

##### client-gen changes

Since client-gen is available for use with 3rd party project, we must ensure all
changes to it are backward compatible. The Apply functions will only be generated
by client-gen if a optional flag is set.

The Apply functions will be included for all built-in types. Strictly speaking
this can be considered a breaking change to the generated client interface, but
adding functions to interfaces is a change we have made in the past, and developers
that have alternate implementations of the interface will usually get a compiler
error in this case, which is relatively trivial to resolve

### Interoperability with structured and unstructured types

For "structs with pointers", json.Marshal, json.Unmarshal and conversions to and

> from unstructured work the same as with go structs.

Reviewer 3: I was thinking about conversion of objects from existing clients/structure to new clients/structure and vice versa. One can always go from one to the other by round-tripping through json/yaml. Would this be useful? Do we have a need for more performance (without going through yaml/json)? Do we want to prevent people from going from existing to new, since that would carry required zero-values? Do we want to implement our own version that doesn't carry zero-value (which would possibly lose some information)?

Reviewer 2: I think the builder should at least have some way of viewing what you've built, e.g. maybe it can emit yaml on demand. Good for debugging and logs.

Author: I agree. The existing prototype includes getter functions. I'll mention them in the KEP here (and wherever else it is appropriate) as well.

For "builders", each could implement `MarshalJSON`, `UnmarshalJSON`,
`ToUnstructured` and `FromUnstructured`.

### Test Plan

#### Fuzz-based round-trip testing

All generated types will be populated using the existing Kubernetes type fuzzer
(see pkg/api/testing) and round tripped to/from the go types. This ensures that
all the generated apply configuration types are able to be correctly populated
with the full state of the go types they mirror.

### Integration testing

The Apply function and the generated apply configuration types will be tested
together in test/integration/client/client_test.go.

### e2e testing

> We will migrate the cluste rrole aggregation controller to use apply and verify

Reviewer 2: ```suggestion
We will migrate the cluster role aggregation controller to use apply and verify
```

why this one?

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

client-go's typed clients need a typesafe, programmatic way to make apply
requests.

## Motivation

Currently, the only way to invoke server side apply from client-go is to call
`Patch` with `PatchType.ApplyPatchType` and provide a `[]byte` containing the
YAML or JSON of the apply configuration. This has a couple important
deficiencies:

- It is a gap completeness of the type client, which provides typesafe APIs for
  all other major methods.
- It makes it to easy for developers to make a major, but non-obvious mistake:
  Use the existing go structs to construct an apply configuration, serialize
  it to JSON, and pass it to `Patch`. This can cause zero valued required
  fields being accientally included in the apply configuration resulting

>   in fields being accidentally set to incorrect values and/or fields accidentally

Reviewer 4: nit: typo accidentally being accidentally

  being accidentally being clamed as owned.

Both sig-api-machinery and wg-api-expression agree that this enhancement is
required for server side apply to be promoted to GA.

### Goals

Introduce a typesafe, programmatic way to call server side apply using the typed
client in client-go.

Express all apply configurations in Go that can be expressed in
YAML. Specifically, an apply request must include only fields that are set by
the applier and exclude those not set by the applier.

Validate this enhancement meets the needs of developers:

- An developer not directly involved in this enhancement successfully converts
  a 1st party controller (one in github.com/kubernetes/kubernetes) to use this
  enhancement.
- A representative group of the developer community is made aware of this
  proposed enhancement, is given early access to it via a fork of
  controller-tools with the requisite generators, and is given the opportunity
  to try it out and provide feedback.

### Non-Goals

Enhancements to client-go's dynamic client. The client-go dynamic client already

> supports Apply via Patch, which is adequate for the dynamic client.

Reviewer 4: I agree that it's adequate for GA of the server-side-apply, but I suspect we'll want this fairly soon after we have it for other types.  Perhaps keep it as a non-goal of this kep, but indicate that some nicer mechanism is desired in the future?

Author: Sounds good. I've updated this section.

Protobuf support. Apply does not support protobuf, and it will not be added with
this enhancement.

## Proposal

`Apply` functions should be included in the typed clients generated for
client-go and should accept the apply configuration using a strongly typed
representation, which will need to be generated for this purpose.

### Risks and Mitigations

#### Poor adoption

Risk: Developers adoption is poor, either because the aesthetics/ergonomics are

> not to their liking or the functinality is insufficient to do what they need to

Reviewer 4: nit: functionality

do with it. This could lead to (a) poor server side apply adoption, and/or (b)
developers building alternate solutions.

Mitigation: We are working with the kubebuilder community to
get hands on feedback from developers to guide our design decisions around
aesthetics/ergonomics with a goal of having both client-go and kubebuilder
take an aligned approach to adding apply to clients in a typesafe way.

## Design Details

### Apply functions

The client-go typed clients will be extended to include Apply functions, e.g.:

```go

> func (c *deployments) Apply(ctx Context, fieldManager string, deployment *appsv1apply.Deployment, opts metav1.ApplyOptions, subresources ...string) (*Deployment, error)

Reviewer 4: instead of modelling based on patch, perhaps we should model based on `Update` which is more familiar in our controllers and workflow.  That would give us `Apply` and `ApplyStatus` which I would expect to translate into the controllers quite well.

Author: Oh right. I fully agree. I'll drop subresources and add `ApplyStatus` and `ApplyScale` where applicable.

```

`ApplyOptions` will be added to metav1 even though `PatchOptions` will continue
to be used over the wire. This will make it obvious in the Apply function

> signature that `fieldManager` is required.

Reviewer 4: I don't think people currently know that field manager is a thing.  The default usually works (I think).  Are we sure that we want to require it as a separate field?  My memory is that the other options structs include it as well, so I would be biased towards consistency if we have a default that usually works.

Author: I've been in a couple discussions that led me to believe that a default can be unsafe.

One is that we some controllers that have multiple independent code paths. E.g. some code path that only updates field X and another that only updates field Y. If a controller like that were naively converted to use apply, it would work fine if it used a distinct field manager for each code path, but would break if it only used a single field manager.

This should probably be put in a `<UNRESOLVED` for now though since there were a couple other open questions about how to handle field manager. I'll do that and add some notes explaining the known issues.

```go

> type ApplyOptions struct {

Reviewer 4: you made me look :)  PatchOptions has a FieldManager today.  Why isn't that reflected here?

Author: It's deliberate. The idea here is to make fieldManger required instead of optional. The way we're proposing doing that is by removing it from the Options and adding it to the Apply function as a parameter.

This is related to the whole "should field manager have a default" question.

```

<<[UNRESOLVED @jpbetz ]>> 
Should the Force field also be made a required field like fieldManager is?
@lavalamp pointed out that user things typically want false but
controller-things typically want true, so false isn't a great default.

Also, can we structure this so that we don't have so many positional params?
Moving two params from options to the top level positional params results in a
longer function signature. Alternative proposed by @lavalamp was to do something
more like this (ordering or parameters and function names TBD):

```go
client.Deployments().BeginApply("field-manager-name", options).Body(
  appsv1apply.Deployment().
    SetObjectMeta(&metav1apply.ObjectMeta()). ...
).Apply(ctx)
```

<<[/UNRESOLVED]>>

Apply will combine the `fieldManager` argument with `ApplyOptions` to create the
`PatchOptions`.

Each apply call will be required to provide a fieldmanager name. We will not
provide a a way for the fieldmanager name to be set for the entire
clientset. The main reason for this is that if a client has multiple code paths
where it makes apply requests to the same object, but with different field sets,

> they must use different field manager names. If they use the same field manager

Reviewer 4: could this be satisfied with a "normal" helper for construction, `NewApplyOptions(fieldManager, force)` and keep the symmetric client methods?

Author: This is a good alternative. I hadn't thought of it. Adding it to the `<<UNRESOLVED` I've got for this whole field manager topic.

problems in practice, there are some that clearly would, e.g. ContainerPort.

Because of this we cannot use the existing go structs to represent apply
configurations.

<<[UNRESOLVED @jpbetz @jennybuckley ]>> 
Finalize which alternative to use based on developer feedback. See the
[Alternatives](#alternatives) for a complete list, but are currently focusing on
the two below alternatives. We are working with the Kubebuilder community to
gather feedback on what developers prefer.
<<[/UNRESOLVED]>>

#### Alternative 1: Generated structs where all fields are pointers

Example usage:

```go
import (
  ptr "k8s.io/utils/pointer"
)

&appsv1apply.Deployment{
  Name: ptr.StringPtr("nginx-deployment"),
  Spec: &appsv1apply.DeploymentSpec{
    Replicas: ptr.Int32Ptr(0),
    Template: &v1apply.PodTemplate{
      Spec: &v1apply.PodSpec{
        Containers: []v1.Containers{
          {
            Name: ptr.StringPtr("nginx"),
            Image: ptr.StringPtr("nginx:latest"),
          },
        }
      },
    },
  },
}
```

For this approach, developers need to use a library like
https://github.com/kubernetes/utils/blob/master/pointer/pointer.go
to inline primitive literals.

> #### Alternative 2: Generated "builders"

Reviewer 4: I like this option better, but I'm fine with both being created for comparison if you like.

Author: I expect this to be the most contentious part of the KEP, I started firmly in the "builders" (alternative 2) camp, now I'm a bit more on the fence, but we do have working implementations of both we are trying out: kubernetes/kubernetes#95988 (struct with pointers) and https://github.com/jpbetz/kubernetes/tree/apply-client-go-builders (builders).

Some of the feedback we have heard from the community:

- "structs with pointers" feels more go idiomatic and more closely aligned with
  the go structs used for Kubernetes types both for builtin types and by
  Kubebuilder.
- It's nice how "builders" are clearly visually distinct from the main go types.
- Having to use a utility library to wrap literal values as pointers for the
  "structs with pointers" is not a big deal. I'm already familiar
  with having to do this in go and once I learn use a utility library for it
  I'm all set.
- The "builders" are awkward to use in an IDE. I felt like I was fighting with
  my IDE to get chain function calls and organize them hierarchally as expected
  by this approach.

TODO: We are providing the developer community with a fork of controller-tools
that will allow them to generate apply configuration types that have both the
alternatives. Our goal is to get feedback from developers after they try out the
generated apply configurations and use that to inform our decision.

While it is possible to generate types that have both the pointer fields exposed

> and the builder functions, we would prefer to provide clear guidance to the

Reviewer 4: I don't know that I'm sold on avoiding the option for each, but I wouldn't let that stop this from proceeding.
